### ISSUE TITLE
Improve diagnostics for variables in nested classes invalidly using type parameters bound in outer scopes

### ISSUE THREAD

**Issue by rzwq** [astral-sh/ty#3562]
### Summary

I encountered a problem where **ty** does not see the generics of an external class (the problem occurred when writing a class using **`sqlalchemy`**, the code below is simplified)

```python
from typing import Optional

class A[T]:
    class B:
        data: Optional[T] = None
```

---

VS Code extension:

<img width="2386" height="376" alt="Image" src="https://github.com/user-attachments/assets/fba16df5-938e-4377-8f36-aacda057ca08" />

Manually:

<img width="1494" height="242" alt="Image" src="https://github.com/user-attachments/assets/2a0684a2-33b4-49e9-93b5-b6c29650163f" />

---

Same problem with **mypy**; **pyright** checks successful

### Version

0.0.39

**Comment by jonathandung** [comment]
python/mypy#21210

**Comment by AlexWaygood** [comment]
ty is working correctly here; this seems like a bug in pyright if pyright allows this. The typing spec [states](https://typing.python.org/en/latest/spec/generics.html#scoping-rules-for-type-variables):

> A generic class nested in another generic class cannot use the same type variables. The scope of the type variables of the outer class doesn’t cover the inner one

And this is enforced in the conformance suite [here](https://github.com/python/typing/blob/e9fccc9dbbd8f1e8b24b4f88911c3d3155059e2a/conformance/tests/generics_scoping.py#L81-L89), with this assertion:

```py
class Outer(Generic[T]):
    class AlsoBad:
        x: list[T]  # E
```

However, our error message here currently seems quite confusing. Looking at the code, you would think that the type variable _is_ bound to an outer generic class. It would be much better if could add a subdiagnostic telling the user that type variables from outer scopes are not valid in nested classes.

I'll repurpose this issue to track improving our diagnostics here.
